security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthenticationManager.scala

```diff
@@ -0,0 +1,107 @@
+package pl.touk.nussknacker.ui.security.api
+
+import akka.http.scaladsl.server.{AuthorizationFailedRejection, Directive0, Directive1, RejectionHandler}
+import akka.http.scaladsl.server.Directives.{handleRejections, optionalHeaderValueByName, provide, reject}
+import cats.data.OptionT
+import pl.touk.nussknacker.security.AuthCredentials.{
+  ImpersonatedAuthCredentials,
+  NoCredentialsProvided,
+  PassedAuthCredentials
+}
+import pl.touk.nussknacker.security.{AuthCredentials, ImpersonatedUserIdentity}
+import pl.touk.nussknacker.ui.security.api.AuthenticationManager.{
+  impersonateHeaderName,
+  impersonationHeaderEndpointInput,
+  mappedAuthenticationEndpointInput
+}
+import sttp.tapir._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthenticationManager(authenticationResources: AuthenticationResources)(implicit ec: ExecutionContext) {
+
+  lazy val authenticationRules: List[AuthenticationConfiguration.ConfigRule] =
+    authenticationResources.configuration.rules
+  lazy val isAdminImpersonationPossible: Boolean = authenticationResources.configuration.isAdminImpersonationPossible
+
+  def authenticate(): Directive1[AuthenticatedUser] = authenticationResources.getAnonymousRole match {
+    case Some(role) =>
+      authenticationResources.authenticate().optional.flatMap {
+        case Some(user) => provideOrImpersonateUser(user)
+        case None => handleAuthorizationFailedRejection.tmap(_ => AuthenticatedUser.createAnonymousUser(Set(role)))
+      }
+    case None => authenticationResources.authenticate().flatMap(provideOrImpersonateUser)
+  }
+
+  def authenticate(authCredentials: AuthCredentials): Future[Option[AuthenticatedUser]] = authCredentials match {
+    case passedCredentials @ PassedAuthCredentials(_) => authenticationResources.authenticate(passedCredentials)
+    case NoCredentialsProvided                        => authenticateWithAnonymousAccess()
+    case ImpersonatedAuthCredentials(impersonatingUserAuthCredentials, impersonatedUserIdentity) =>
+      authenticateWithImpersonation(impersonatingUserAuthCredentials, impersonatedUserIdentity)
+  }
+
+  def authenticationEndpointInput(): EndpointInput[AuthCredentials] =
+    authenticationResources
+      .authenticationMethod()
+      .and(impersonationHeaderEndpointInput)
```

Comment:
it would be great if we could write it like that:
```scala
 authenticationResources
      .authenticationMethod()
      .withPossibleImpersonation()
```

and it could be implemented like this:
```
  
  implicit class ImpersonationConsideringInputEndpoint(underlying: EndpointInput[Option[String]]) {
    def withPossibleImpersonation(): EndpointInput[AuthCredentials] = {
      underlying
        .and(impersonationHeaderEndpointInput)
        .map(mappedAuthenticationEndpointInput)
    }
    
    private def impersonationHeaderEndpointInput = ???
    private def mappedAuthenticationEndpointInput = ???
```

We could retuse this extension class in tests too (see NuDesignerApiAvailableToExposeYamlSpec)

Reply:
Done, I introduced the implcicit class.